src/System/IO/Resource/Linear/Internal.hs

```diff
@@ -98,101 +113,87 @@ instance Control.Monad RIO where
 openBinaryFile :: FilePath -> System.IOMode -> RIO Handle
-openBinaryFile path mode = Control.do
-  h <-
-    unsafeAcquire
-      (Linear.fromSystemIOU $ System.openFile path mode)
-      (\h -> Linear.fromSystemIO $ System.hClose h)
-  Control.pure $ Handle h
+openBinaryFile path mode =
+  unsafeAcquire
+    (Linear.fromSystemIOU $ System.openFile path mode)
+    (\h -> Linear.fromSystemIO $ System.hClose h)
 
+-- | Specialised alias for 'release'
 hClose :: Handle %1 -> RIO ()
-hClose (Handle h) = unsafeRelease h
+hClose = release
 
 hIsEOF :: Handle %1 -> RIO (Ur Bool, Handle)
-hIsEOF = coerce (unsafeFromSystemIOResource System.hIsEOF)
+hIsEOF = unsafeFromSystemIOResource System.hIsEOF
 
 hGetChar :: Handle %1 -> RIO (Ur Char, Handle)
-hGetChar = coerce (unsafeFromSystemIOResource System.hGetChar)
+hGetChar = unsafeFromSystemIOResource System.hGetChar
 
 hPutChar :: Handle %1 -> Char -> RIO Handle
-hPutChar h c = flipHPutChar c h -- needs a multiplicity polymorphic flip
-  where
-    flipHPutChar :: Char -> Handle %1 -> RIO Handle
-    flipHPutChar c =
-      coerce (unsafeFromSystemIOResource_ (\h' -> System.hPutChar h' c))
+hPutChar h c = unsafeFromSystemIOResource_ (\h' -> System.hPutChar h' c) h
 
 hGetLine :: Handle %1 -> RIO (Ur Text, Handle)
-hGetLine = coerce (unsafeFromSystemIOResource Text.hGetLine)
+hGetLine = unsafeFromSystemIOResource Text.hGetLine
 
 hPutStr :: Handle %1 -> Text -> RIO Handle
-hPutStr h s = flipHPutStr s h -- needs a multiplicity polymorphic flip
-  where
-    flipHPutStr :: Text -> Handle %1 -> RIO Handle
-    flipHPutStr s =
-      coerce (unsafeFromSystemIOResource_ (\h' -> Text.hPutStr h' s))
+hPutStr h s = unsafeFromSystemIOResource_ (\h' -> Text.hPutStr h' s) h
 
 hPutStrLn :: Handle %1 -> Text -> RIO Handle
-hPutStrLn h s = flipHPutStrLn s h -- needs a multiplicity polymorphic flip
-  where
-    flipHPutStrLn :: Text -> Handle %1 -> RIO Handle
-    flipHPutStrLn s =
-      coerce (unsafeFromSystemIOResource_ (\h' -> Text.hPutStrLn h' s))
+hPutStrLn h s = unsafeFromSystemIOResource_ (\h' -> Text.hPutStrLn h' s) h
 
 -- | See @System.IO.'System.IO.hSeek'@.
 --
 -- @since 0.3.0
 hSeek :: Handle %1 -> System.SeekMode -> Integer -> RIO Handle
-hSeek h mode i = coerce hSeek' h
-  where
-    hSeek' :: Handle %1 -> RIO Handle
-    hSeek' =
-      coerce (unsafeFromSystemIOResource_ (\h' -> System.hSeek h' mode i))
+hSeek h mode i = unsafeFromSystemIOResource_ (\h' -> System.hSeek h' mode i) h
 
 -- | See @System.IO.'System.IO.hTell'@.
 --
 -- @since 0.3.0
 hTell :: Handle %1 -> RIO (Ur Integer, Handle)
-hTell = coerce (unsafeFromSystemIOResource System.hTell)
+hTell = unsafeFromSystemIOResource System.hTell
 
 -- new-resources
 
 -- | The type of system resources.  To create and use resources, you need to
 -- use the API since the constructor is not released.
-data UnsafeResource a where
-  UnsafeResource :: Int -> a -> UnsafeResource a
+data Resource a where
+  UnsafeResource :: Int -> a -> Resource a
```

Comment:
If the constructor is not exported, does that mean it's safe to rename?

Reply:
It is safe to rename. But I like how the name emphasises that accessing this constructor is very unsafe and needs to be done with care. What would you like to see instead?

Reply:
I think that makes sense. Just checking.